Source/AliveLib/Scrab.cpp

```diff
@@ -627,8 +685,432 @@ __int16 Scrab::AI_Patrol_0_4AA630()
+        {
+            MusicController::sub_47FD60(8, this, 0, 0);
+        }
+        switch (field_11C_sub_state)
+        {
+        case AI_ChasingEnemy::eState1_UNKNOWN_0:
+            field_150 = sGnFrame_5C1B84 + field_128_attack_delay;
+            return AI_ChasingEnemy::eState1_Idle_1;
+        case AI_ChasingEnemy::eState1_Idle_1:
+            if (!field_106_current_motion)
+            {
+                field_194 = LastSpeak_4A56F0();
+                if (field_1A2 < field_1A0
+                    && LastSpeak_4A56F0() == GameSpeakEvents::eUnknown_53
+                    || LastSpeak_4A56F0() == GameSpeakEvents::eUnknown_54)
+                {
+                    return AI_ChasingEnemy::eState1_UNKNOWN_15;
+                }
+            }
+            if (!vIsFacingMe_4254A0(pObj))
+            {
+                if (gMap_5C3030.Is_Point_In_Current_Camera_4810D0(field_C2_lvl_number, field_C0_path_number, field_B8_xpos, field_BC_ypos, 0))
+                {
+                    field_108_next_motion = eScrabMotions::M_Turn_3_4A91A0;
+                }
+                else
+                {
+                    field_106_current_motion = eScrabMotions::M_Turn_3_4A91A0;
+                    field_108_next_motion = -1;
+                    MapFollowMe_408D10(TRUE);
+                }
+                return AI_ChasingEnemy::eState1_Turning_3;
+            }
+            if (vIsObjNearby_4253B0(ScaleToGridSize_4498B0(field_CC_sprite_scale) / FP_FromInteger(5), pObj)
+                && vOnSameYLevel_425520(pObj))
+            {
+                if (!vIsObjNearby_4253B0(ScaleToGridSize_4498B0(field_CC_sprite_scale), pObj))
+                {
+                    field_108_next_motion = eScrabMotions::M_AttackLunge_37_4AA0B0;
+                    return AI_ChasingEnemy::eState1_Attacking_8;
+                }
+                field_108_next_motion = eScrabMotions::M_LegKick_38_4AA120;
+                return AI_ChasingEnemy::eState1_Attacking_8;
+            }
+            else if (CanSeeAbe_4A51A0(pObj) && static_cast<int>(sGnFrame_5C1B84) >= field_150)
+            {
+                if (Handle_SlamDoor_or_EnemyStopper_4A4830(field_C4_velx, 0))
+                {
+                    return field_11C_sub_state;
+                }
+                field_108_next_motion = eScrabMotions::M_Run_2_4A89C0;
+                return AI_ChasingEnemy::eState1_Running_2;
+            }
+            else
+            {
+                if (Math_NextRandom() >= 16u || static_cast<int>(sGnFrame_5C1B84 - field_154) <= 60)
+                {
+                    return AI_ChasingEnemy::eState1_Idle_1;
+                }
+                field_108_next_motion = eScrabMotions::M_Shriek_30_4A9EA0;
+                field_154 = sGnFrame_5C1B84;
+                return AI_ChasingEnemy::eState1_Shriek_14;
+            }
+        case AI_ChasingEnemy::eState1_Running_2:
+            field_194 = LastSpeak_4A56F0();

```

Comment:
I think all of case 2 should go into its own helper function because its huge and will become easier to refactor some parts of it